services/service.ts

```diff
@@ -0,0 +1,341 @@
+import fs from 'fs';
+import { createHash } from 'crypto';
+import fetch from 'node-fetch';
+import path from 'path';
+import child_process from 'child_process';
+
+// @ts-ignore
+import { Promise } from 'bluebird';
+import progress from 'progress-stream';
+import { Service, Message } from 'webos-service';
+
+import serviceInfo from './services.json';
+import { asyncAccess, asyncExecFile, asyncPipeline, asyncUnlink, asyncWriteFile } from './adapter';
+import { makeError, makeSuccess } from './protocol';
+import ServiceRemote from './webos-service-remote';
+
+const kHomebrewChannelPackageId = serviceInfo.services[0].name;
+
+const service = new Service(serviceInfo.id);
+const serviceRemote = new ServiceRemote(service);
+
+function runningAsRoot() {
+  return process.getuid() === 0;
+}
+
+function getInstallerService(): Service {
+  if (runningAsRoot()) {
+    return service;
+  }
+  return serviceRemote as Service;
+}
+
+// Maps internal setting field name with filesystem flag name.
+type FlagName = string;
+const availableFlags = {
+  telnetDisabled: 'webosbrew_telnet_disabled',
+  failsafe: 'webosbrew_failsafe',
+  sshdEnabled: 'webosbrew_sshd_enabled',
+  blockUpdates: 'webosbrew_block_updates',
+} as Record<string, FlagName>;
+
+function asyncCall<T extends Record<string, any>>(srv: Service, uri: string, args: Record<string, any>): Promise<T> {
+  return new Promise((resolve, reject) => {
+    srv.call(uri, args, ({ payload }) => {
+      if (payload.returnValue) {
+        resolve(payload as T);
+      } else {
+        reject(payload);
+      }
+    });
+  });
+}
+
+function createToast(message: string): Promise<Record<string, any>> {
+  return asyncCall(service, 'luna://com.webos.notification/createToast', {
+    sourceId: serviceInfo.id,
+    message,
+  });
+}
+
+/**
+ * Generates local file checksum.
+ */
+async function hashFile(filePath: string, algorithm: string): Promise<string> {
+  const download = fs.createReadStream(filePath);
+  const hash = createHash(algorithm, { encoding: 'hex' });
+  await asyncPipeline(download, hash);
+  return hash.read();
+}
+
+/**
+ * Elevates a package by name.
+ */
+async function elevateService(pkg: string) {
+  if (runningAsRoot()) {
+    console.info('Elevating service...');
+    await asyncExecFile(path.join(__dirname, 'elevate-service'), [pkg]);
+    console.info('Finished, dying as soon as possible.');
+  }
+  // @ts-ignore
+  service.activityManager.idleTimeout = 1;
```

Comment:
I think this and `dying...` comment above should end up in the callee instead if this is meant to be generic - we lower the `idleTimeout` to quit when no more messages are in flight (since `appInstallService` doesn't touch running services when upgrading), but this is not really required if you just want to elevate some service.

Reply:
You're completely right. Fixed.